### PR TITLE
suchThat() Generator

### PR DESCRIPTION
**Pull request by giorgiosironi**
The example demonstrates how it can be useful to filter a Generator in a way that can be composed further, implementing something that is impossible by adding only a `when()` at the `$this->forAll()` top level.

The reference is always https://github.com/clojure/test.check/blob/master/doc/intro.md but I would like to call this `filter` in reference to http://php.net/array_filter (but with the `callable, generator` order).

### CODE REVIEW

File: src/Eris/Generator/SuchThatGenerator.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        $this->maximumAttempts = 10;
+    }
+
+    public function __invoke($size)
+    {
+        $value = $this->generator->__invoke($size);
+        $attempts = 0;
+        while (!$this->predicate($value)) {
+            if ($attempts >= $this->maximumAttempts) {
+                throw new LogicException("Tried to satisfy predicate $attempts times, but could not generate a good value. Last one was: " . $value);
```
Comment (MirkoBonadei) [inline src/Eris/Generator/SuchThatGenerator.php]: Can we provide a better error message here? We can keep this one but maybe we can add something useful to solve the problem, something like "You should try to improve your generator".

**Comment by MirkoBonadei** [comment]
:+1: 

I don't have a strong opinion on the name of this generator. Basically I would keep the property based testing vocabulary (since there are no books on PBT in PHP a programmer may feel lost looking for a suchThat after reading the book on ScalaCheck), but I understand also yor effort of making the PHP developer feel like she/he is at home.

**Comment by giorgiosironi** [comment]
I would call it `filter` and `FilterGenerator` but leave in also `suchThat` as an alias.

**Comment by giorgiosironi** [comment]
Or the converse, let's see that the other Generators do

**Pull request merged**